util: add prototype support for function inspection

BridgeAR opened the pull request:
<s>This prevents function names being printed if the function is part
of an object and the property it's attached to has the same name as
the function.</s>

This adds prototype support for functions when using `util.inspect()` and marks anonymous functions explicitly as such.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mscdex commented:
Doesn't this make them appear identical to anonymous functions now?

BridgeAR commented:
@mscdex that is correct. The same will almost always be assigned and anonymous functions are rare for things to inspect.

What about highlighting anonymous functions with either a different color or by explicitly mentioning that it's anonymous? That way the average would be less verbose and less duplication.

If anyone has some further suggestions how to improve it further by reducing the overhead, that would be great! :-)

BridgeAR commented:
@nodejs/util PTAL

BridgeAR commented:
This could use some reviews.

BridgeAR commented:
@nodejs/util @mscdex @joyeecheung I thought about distinguishing anonymous functions and other functions again and the best I could think of is switching the brackets. We could just use `<Function>` instead of `[Function]`, document the difference and use `[Anonymous function]` for all anonymous functions. That way it's explicit that a function is anonymous and it's possible to see the difference between removed function names in case it's a property.

Any opinions on that?

mscdex commented:
To me it seems like anonymous functions as property values would be more common (especially when considering object literal declarations), so keeping things the way they are now is more ideal IMO.

BridgeAR commented:
@mscdex anonymous functions as property values are normally using the property name as function name:

```
> { a() {}, b: () => {}, c: function() {} }
{ a: [Function: a], b: [Function: b], c: [Function: c] }
```

joyeecheung commented:
Can we use (anonymous function) like how it’s used in the inspector?

BridgeAR commented:
I updated the branch with my suggestions. That is now a somewhat more involved change than before, since I also added null prototype support and made functions more tampering proof besides just marking anonymous functions as such.

@joyeecheung PTAL. Using `(anonymous)` alone would not be sufficient, since there are different types of functions (async and generator ones). If you suggested to change it to: `[Function: (anonymous)]`, that would be a possibility while it would be possible to set that as name with `defineProperty` (even though it would be unlikely).

BridgeAR commented:
The first commit is the original one. I did not change that during the rebase. The first two commits should be squashed later but I thought I keep it around for easier reviews.

joyeecheung commented:
> If you suggested to change it to: [Function: (anonymous)], that would be a possibility while it would be possible to set that as name with defineProperty (even though it would be unlikely).

Yes, that’s what I meant, because the first part is supposed to be the type isn’t it?

I think it’s fine to use it even though the name can be set using defineProperty, Chrome still does this way so this is a format people are familiar with. With the Anonymous prefix and the brackets it’s less obvious without docs.

mscdex commented:
Instead of `[Function: (anonymous)]` you could leave out the `:`, that would avoid ambiguity if someone named their function `(anonymous)`.

BridgeAR commented:
Thanks for your review! I updated it to the suggestion.

addaleax commented:
This is definitely semver-major, btw. We’ve even had issues when the function name display changed during a semver-minor V8 upgrade in the past.

BridgeAR commented:
@addaleax 
> We’ve even had issues when the function name display changed during a semver-minor V8 upgrade in the past.

But that was about `String(function)`, not about `util.inspect()` which does something very different.

> This is definitely semver-major

I agree due to the change of the brackets. Without that, I don't think that it would have to be semver-major.

> I don’t think people get a good intuition for what the difference between <…> and […] mean, respectively, when inspecting values.

Since I also added `anonymous` in case a function is anonymous, I guess we could also stick to the square brackets as before. I was reluctant to change the brackets first due to the non-intuitive character but I think as soon as people saw the difference a couple of times, it would be something good. Especially to differentiate it with older Node.js versions (that is my main concern about keeping the square brackets).
Do you have a suggestion how to improve it further? I personally think this change became pretty decent.

addaleax commented:
> > We’ve even had issues when the function name display changed during a semver-minor V8 upgrade in the past.

> But that was about `String(function)`, not about `util.inspect()` which does something very different.

No, I was talking about `util.inspect()` – iirc, when V8 started inferring function names for properties, it showed up in `util.inspect()` output that people were testing against.

> > This is definitely semver-major

> I agree due to the change of the brackets. Without that, I don't think that it would have to be semver-major.

I think this is semver-major either way, it changes `util.inspect()` output rather significantly (the number of tests that had to be changed for that is a pretty good indicator), and while we discourage people from using it programatically, I think we should err on the side of caution. That also gives people a longer time to figure out whether this is a change that improves `util.inspect()` before we release it.

> > I don’t think people get a good intuition for what the difference between <…> and […] mean, respectively, when inspecting values.

> Since I also added `anonymous` in case a function is anonymous, I guess we could also stick to the square brackets as before

I would prefer that, yes.

> Do you have a suggestion how to improve it further? I personally think this change became pretty decent.

I mean, still, in the end it seems like this PR removes redundancy in one way and adds it in another way, and gives up consistency during that? I don’t have strong opinions about this besides that I find the different brackets confusing, so I’d defer to others…

BridgeAR commented:
> No, I was talking about util.inspect() – iirc, when V8 started inferring function names for properties

I just looked through a lot of issues but I could not find anything related. Do you have a reference for it for me?

> the number of tests that had to be changed for that is a pretty good indicator

This is actually a combination of multiple minor changes. Should I maybe just move a few changes out in a separate PR? For example everything besides the `anonymous` part and the property name part?

> > Since I also added anonymous in case a function is anonymous, I guess we could also stick to the square brackets as before

> I would prefer that, yes.

OK, but this will conflict with older Node.js versions in that case. So if we have another solution / idea, I would rather do that.

> I mean, still, in the end it seems like this PR removes redundancy in one way and adds it in another way, and gives up consistency during that? I don’t have strong opinions about this besides that I find the different brackets confusing, so I’d defer to others…

It adds the redundancy for `anonymous` functions but that should be less common than having a function as property name. It is also more explicit in case of anonymous functions what I personally prefer.

------------------

CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1830/ :heavy_check_mark:

addaleax commented:
@BridgeAR 9940766f8a09c2eec06f582cbb0b4883f601be89 is essentially what I was referring to, but I also remember something this impacting outside projects. (I vaguely remember it being one of @bcoe’s projects that was affected by `util.inspect()` changes but I could be wrong).

> It adds the redundancy for `anonymous` functions but that should be less common than having a function as property name. 

Fwiw, I’d think that anonymous function are generally more common than objects with enumerable function properties.

BridgeAR commented:
> https://github.com/nodejs/node/commit/9940766f8a09c2eec06f582cbb0b4883f601be89 is essentially what I was referring to

I understand what change you meant but I can't find any issues or comments about trouble due to this change.

> Fwiw, I’d think that anonymous function are generally more common than objects with enumerable function properties.

I guess in general that might be correct but those are normally not logged, are they? (I can mostly think of functions logged as object value or in case the function is assigned directly to a variable. Both of these cases infer the function name in almost all cases.)

BridgeAR commented:
I completely removed the original intent of reducing the redundancy. This change is now something pretty different from the original part and it would require new reviews (cc @joyeecheung).

The reason to remove that part is that handling it as `[Function]` would conflict with older Node.js versions handling anonymous functions. They should somehow be marked different but there was a -1 for now, so I think it's best to focus on the rest of the PR instead and we could still come back to the rest at another point.

@addaleax do you still consider this semver-major?

bcoe commented:
@addaleax nothing immediately rings a bell.

BridgeAR commented:
Rebased due to conflicts.

BridgeAR commented:
@nodejs/tsc PTAL. This requires another LG from the TSC (I personally don't think this has to be semver-major: CITGM is clean and we handle most changes to `util.inspect` as patch. It's not that likely that this change breaks tests, especially since we warn not to rely programmatically on the output).

thefourtheye commented:
> Instead of [Function: (anonymous)] you could leave out the :, that would avoid ambiguity if someone named their function (anonymous)

Isn't naming a function `(anonymous)` syntactically wrong?

BridgeAR commented:
@thefourtheye 
> Isn't naming a function (anonymous) syntactically wrong?

It's not possible to name a function like that when creating them but it's possible to change the name later on by using `Object.defineProperty`.

BridgeAR commented:
Landed in d0667e814e8be53d329a9c7f4849996c192395c9 :tada:

@addaleax since we try to backport semver-major PRs in non-breaking ways: I'll open a backport for v12 that backports the prototype part but not the anonymous part. Does that sound fine to you?

addaleax commented:
@bridgear Yes, I think so.

targos commented:
@BridgeAR please do :) This is blocking https://github.com/nodejs/node/pull/27522.

BridgeAR commented:
@targos I opened a backport (#27570) but the backport will not resolve that conflict. That has to be done manually.

BridgeAR commented:
I just added the backport for #27522 to the same PR.
